src/Xrefcheck/Scanners/Markdown.hs

```diff
@@ -82,33 +84,73 @@ cataNode f (Node pos ty subs) = f pos ty (cataNode f <$> subs)
 
 -- | Remove nodes with accordance with global `MarkdownConfig` and local
 --   overrides.
-removeIgnored :: Node -> Either Text Node
-removeIgnored = runIdentity . runExceptT . flip evalStateT None . cataNode remove
+removeIgnored :: MonadWriter [ScanError] m => FilePath -> Node -> m Node
+removeIgnored fp = withIgnoreMode . cataNode remove
   where
     remove
-      :: (MonadError Text m, MonadState IgnoreMode m)
+      :: (MonadWriter [ScanError] m, MonadState IgnoreMode m)
       => Maybe PosInfo
       -> NodeType
       -> [m Node]
       -> m Node
     remove pos ty subs = do
       mode <- get
+      let node = Node pos ty []
       case (mode, ty) of
         (Paragraph, PARAGRAPH) -> put None $> defNode
-        (Paragraph, x)         -> throwError (makeError mode (prettyType x) pos)
-        (File, _)              -> throwError (makeError mode "" pos)
-        (Link, LINK {})        -> put None $> defNode
-        (Link, _)              -> Node pos ty <$> sequence subs
-        (None, _) -> do
-          case getIgnoreMode (Node pos ty []) of
-            Just mode' -> put mode' $> defNode
+        (Paragraph, x)         -> do
+          tell . scanError pos . makeError mode $ prettyType x
+          put None
+          Node pos ty <$> sequence subs
+        (File, _)              -> do
+          tell . scanError pos $ makeError mode ""
+          put None
+          Node pos ty <$> sequence subs
+        (Link _, LINK {})      -> put None $> defNode
+        (Link pos', _)         ->
+          case getIgnoreMode node of
+            Just mode'  -> do
+              tell . scanError pos' $ makeError mode ""
+              handleMode node mode'
+            Nothing     -> Node pos ty <$> sequence subs
+        (None, _)              ->
+          case getIgnoreMode node of
+            Just mode' -> handleMode node mode'
             Nothing    -> Node pos ty <$> sequence subs
 
+    handleMode
+      :: (MonadWriter [ScanError] m, MonadState IgnoreMode m)
+      => Node
+      -> IgnoreMode
+      -> m Node
+    handleMode node@(Node pos _ _)= \case
+      None   -> do
+        let unrecognised = fromMaybe ""
+              $ safeHead . drop 1 . words =<< getXrefcheckContent node
+        tell . scanError pos $ makeError None unrecognised
+        put None $> defNode
+      Link _ -> put (Link pos) $> defNode
+      mode'  -> put mode' $> defNode
+
+    scanError :: Maybe PosInfo -> Text -> [ScanError]
+    scanError pos = pure . ScanError (toPosition pos) fp
+
     prettyType :: NodeType -> Text
     prettyType ty =
       let mType = safeHead $ words $ show ty
       in fromMaybe "" mType
 
+    withIgnoreMode
+      :: (MonadWriter [ScanError] m)
+      => StateT IgnoreMode m Node
+      -> m Node
+    withIgnoreMode action = action `runStateT` None >>= \case
+      (node, link@(Link pos)) -> do
```

Comment:
Same here: why does the `Link` case get special treatment?

If the scanning `action` ends with a `Paragraph`/`File` mode, we don't report an error (presumably because it's already been reported _during_ the scanning). But if it ends with a `Link` mode, we report an error - why? Shouldn't it already have been reported during the scanning `action`?

Reply:
For the reasons mentioned above, I thought the link might be deeply nested in other nodes, so we had to report that we reached EOF without finding any links. In fact, we probably need to keep this even if we change the behavior of the `ignore link` annotation so that we can report that there is no link after the annotation if the annotation is used at the end of file. Also, right now, if we use `ignore paragraph` on the last line of the file, nothing is reported.

Reply:
Okay that makes sense!

> Also, right now, if we use ignore paragraph on the last line of the file, nothing is reported, we probably need to report that as well.

Yup, that's true. More generally, the ignore mode must be set to `None` when we reach the end of the file.

Reply:
A non-`None` mode implies that there was an annotation in the file that didn't match any node.